feat: resource link component for semiotics [WD-15623]

mas-who opened the pull request:
## Done

- Added a chip component for displaying lxd resources
- Added custom icons that we cannot get from vanilla
- Mock implementation for instance creation toast notifications

## QA

1. Run the LXD-UI:
    - On the demo server via the link posted by @webteam-app below. This is only available for PRs created by collaborators of the repo. Ask @mas-who or @edlerd for access.
    - With a local copy of this branch, [build and run as described in the docs](../CONTRIBUTING.md#setting-up-for-development).
2. Perform the following QA steps:
    - TBC

## Screenshots
![Screenshot from 2024-10-10 11-47-52](https://github.com/user-attachments/assets/7e643e8e-eb32-41bd-af5a-ce14731f3088)

webteam-app commented:
### [<img src='https://assets.ubuntu.com/v1/6baef514-ubuntu-circle-of-friends-large.svg' height=32 width=32> Demo</img>](https://lxd-ui-946.demos.haus)
### [<img src='https://assets.ubuntu.com/v1/e512b0e2-jenkins.svg' height=32 width=32> Jenkins </img>](https://jenkins.canonical.com/webteam/job/start-demo/19300)
### [<img src='https://assets.ubuntu.com/v1/7144ec6d-logo-jaas-icon.svg' height=32 width=32> demos.haus </img>](https://demos.haus)

edlerd commented on src/components/ResourceLink.tsx:
```diff
@@ -0,0 +1,60 @@
+  const icon = resourceIconMap[type];
+  const navigate = useNavigate();
+  return (
+    <button className="p-chip is-inline" onClick={() => navigate(to)}>
```
I think button is the wrong component to use and this should be a `Link` to be semantically correct. Then the browser would display the target url correctly and you can do things like open in a new tab.

edlerd commented on src/sass/_pattern_icon.scss:
```diff
@@ -2,11 +2,83 @@
   background-image: url("data:image/svg+xml,%3Csvg viewBox='0 0 16 16' width='16px' height='16px' fill='none' xmlns='http://www.w3.org/2000/svg'%3E%3Cpath d='M 0 7.743 L 6.742 14.485 L 7.899 13.329 L 2.311 7.743 L 7.899 2.157 L 6.742 1 L 0 7.743 Z M 7.899 7.743 L 14.642 14.485 L 15.797 13.329 L 10.21 7.743 L 15.797 2.157 L 14.642 1 L 7.899 7.743 Z' fill='%23FFF' style=''/%3E%3C/svg%3E");
 }
 
+@mixin lxdui-icon-lxd-container() {
+  background-image: url("data:image/svg+xml,%3Csvg viewBox='0 0 16 16' width='16px' height='16px' fill='none' xmlns='http://www.w3.org/2000/svg'%3E%3Cpath fill-rule='evenodd' clip-rule='evenodd' d='M13.5 10.9901V5.84556L11.5 6.96094V10.0536L9.75 11.0536L8.74998 11.625V13.619L10.7736 12.499L13.5 10.9901ZM7.24998 11.625V13.6191L5.22632 12.499L2.49994 10.9901V5.84554L4.5 6.96096V10.0536L6.25 11.0536 ...
+}
+
+@mixin lxdui-icon-snapshot() {
+  background-image: url("data:image/svg+xml,%3Csvg viewBox='0 0 16 16' width='16px' height='16px' fill='none' xmlns='http://www.w3.org/2000/svg'%3E%3Cpath%20fill-rule%3D%22evenodd%22%20clip-rule%3D%22evenodd%22%20d%3D%22M5.58127%204.50702L5.22464%205.5H2.5V12.5H13.5V5.5H10.7754L10.4187%204.50702L10.418%204.50532C10.4175%204.50424%2010.4166%204.50246%2010.4152%204.49999L10.1295%204H5.87048L5.58477 ...
+}
+
+@mixin lxdui-icon-repository() {
+  background-image: url("data:image/svg+xml,%3Csvg viewBox='0 0 16 16' width='16px' height='16px' fill='none' xmlns='http://www.w3.org/2000/svg'%3E%3Cpath%20fill-rule%3D%22evenodd%22%20clip-rule%3D%22evenodd%22%20d%3D%22M9.22565%201.18947L14.4505%204.12846C14.9319%204.39923%2015.1026%205.00894%2014.8318%205.4903C14.7421%205.64993%2014.6102%205.78183%2014.4505%205.87161L9.22565%208.81061C8.46462%2 ...
+}
+
+@mixin lxdui-icon-folder() {
+  background-image: url("data:image/svg+xml,%3Csvg viewBox='0 0 16 16' width='16px' height='16px' fill='none' xmlns='http://www.w3.org/2000/svg'%3E%3Cpath%20fill-rule%3D%22evenodd%22%20clip-rule%3D%22evenodd%22%20d%3D%22M6%205.46408V3.96408V3.60693H2.5V5.46408V6.60291V8.66667L2.94235%207.09388C3.12429%206.44696%203.7143%206%204.38632%206H11.5V5.46408H7.5H6ZM13%206H14.0216C15.0862%206%2015.812%207 ...
+}
+
+@mixin lxdui-icon-single-host() {
+  background-image: url("data:image/svg+xml,%3Csvg viewBox='0 0 16 16' width='16px' height='16px' fill='none' xmlns='http://www.w3.org/2000/svg'%3E%3Cpath%20fill-rule%3D%22evenodd%22%20clip-rule%3D%22evenodd%22%20d%3D%22M13.9658%204.55568L7.99998%201.11133L2.03418%204.55568V11.4444L7.99998%2014.8888L13.9658%2011.4444V4.55568ZM3.53398%205.42104L7.99898%202.84304L12.465%205.42104V10.578L7.99898%201 ...
+}
+
+@mixin lxdui-icon-iso() {
+  background-image: url("data:image/svg+xml,%3Csvg viewBox='0 0 16 16' width='16px' height='16px' fill='none' xmlns='http://www.w3.org/2000/svg'%3E%3Cpath%20fill-rule%3D%22evenodd%22%20clip-rule%3D%22evenodd%22%20d%3D%22M12.9204%205.53979C13.2913%206.28006%2013.5%207.11566%2013.5%208C13.5%2011.0376%2011.0376%2013.5%208%2013.5C7.11566%2013.5%206.28006%2013.2913%205.53979%2012.9204L7.10538%209.7892 ...
+}
+
+@mixin lxdui-icon-image() {
+  background-image: url("data:image/svg+xml,%3Csvg viewBox='0 0 16 16' width='16px' height='16px' fill='none' xmlns='http://www.w3.org/2000/svg'%3E%3Cpath%20fill-rule%3D%22evenodd%22%20clip-rule%3D%22evenodd%22%20d%3D%22M13.5%208C13.5%2011.0376%2011.0376%2013.5%208%2013.5C4.96243%2013.5%202.5%2011.0376%202.5%208C2.5%204.96243%204.96243%202.5%208%202.5C11.0376%202.5%2013.5%204.96243%2013.5%208ZM15 ...
+}
+
+@mixin lxdui-icon-certificate() {
+  background-image: url("data:image/svg+xml,%3Csvg viewBox='0 0 16 16' width='16px' height='16px' fill='none' xmlns='http://www.w3.org/2000/svg'%3E%3Cpath%20fill-rule%3D%22evenodd%22%20clip-rule%3D%22evenodd%22%20d%3D%22M13%202C14.1046%202%2015%202.89543%2015%204V11C15%2012.1046%2014.1046%2013%2013%2013H12V11.5H13C13.2455%2011.5%2013.4496%2011.3231%2013.4919%2011.0899L13.5%2011V4C13.5%203.75454%2 ...
+}
+
```
We should update vanilla first, so we can import those instead of defining them here.

mas-who replied:
still waiting for e2e to pass on #947

edlerd commented on src/components/ResourceLink.tsx:
Can you add 1 or 2 more use cases in pages, maybe with the different types, so it is easier to see the component in action without having to create a new instance?

mas-who replied:
I added the new component to the instance detail overview page. I think with the new design from @piperdeck, where the border is removed from the chip, it looks decent. Side note this is after discussion between piper and lyubo. I'm not too sure about the status dots for storage pools (and volumes) though.

![Screenshot from 2024-10-11 11-59-08](https://github.com/user-attachments/assets/4bfbaeb1-27b7-43f5-9044-611b6776b748)

mas-who commented:
@edlerd @piperdeck I have applied the chip design across all floating notifications. Also implemented them on the instance detail overview page. Let's use the PR to help with our discussion later.

PS: also updated the side nav icons

edlerd commented on src/components/Navigation.tsx:
i.e. the instance list https://lxd-ui-946.demos.haus/ui/project/No-Root-Disk-Project/instances should have the same icon in the empty state as the navigation icon.

mas-who replied:
Ahh yes, I think safe bet is to do a global search for all `<Icon />` invocations in the codebase and see if there are any other missed scenarios

edlerd commented on src/pages/cluster/ClusterGroupForm.tsx:
```diff
@@ -85,7 +86,17 @@ const ClusterGroupForm: FC<Props> = ({ group }) => {
           const verb = group ? "saved" : "created";
           navigate(`/ui/cluster/group/${values.name}`);
-          toastNotify.success(`Cluster group ${values.name} ${verb}.`);
+          toastNotify.success(
+            <>
+              Cluster group{" "}
+              <ResourceLink
+                type="cluster-member"
+                value={values.name}
+                to={`/ui/cluster/group/${values.name}`}
+              />{" "}
+              {verb}.
+            </>,
+          );
```
I think we should always keep the resource type as part of the copy, like in this example. That makes it easy to learn the icons and reduces the cognitive load, i.e. when browsing notifications. Currently, you'd need to understand what each icon stands for to process a notification. Putting the type name as an explicit part of the copy makes it much easier to process.

mas-who replied:
This is another design decision we need to align on with @piperdeck. I believe the purpose of using icons is to eliminate the need for resource type text in notifications. In last week’s discussion, we agreed that only storage pools and custom volumes should include their types in the notification copy, as we don’t yet have representative icons for them. Before proceeding with this change, I’d like to confirm this approach, as it will require updates to all the e2e tests.

edlerd replied:
Sure, lets discuss with piper. When I was playing with this branch I found it hard to digest the notifications, hence my suggestion.

> Before proceeding with this change, I’d like to confirm this approach, as it will require updates to all the e2e tests.

I thought this would actually make the e2e test easier, because we had this in the past. So with this suggestion, the CI suite would need fewer adjustments.

mas-who replied:
Aaah right I already made adjustments to all the e2e tests :sweat_smile: but it's easy enough to reset the commit and remove those changes, but let's first confirm the change with piper. I do agree that it's additional learning if we remove the resource type copy text from the notifications

mas-who replied:
@edlerd after discussing with @piperdeck I've added the resource type back into all notifications. Also reverted all the e2e test changes.

mas-who replied:
Related to this [comment](https://github.com/canonical/lxd-ui/pull/952#issuecomment-2414631370) from #952 . I have adjusted the instance side navigation tab and empty state to use the `pod` icon.

edlerd commented on src/pages/cluster/ClusterGroupForm.tsx:
```diff
@@ -85,7 +86,17 @@ const ClusterGroupForm: FC<Props> = ({ group }) => {
-          toastNotify.success(`Cluster group ${values.name} ${verb}.`);
+          toastNotify.success(
+            <>
+              Cluster group{" "}
+              <ResourceLink
+                type="cluster-member"
+                value={values.name}
+                to={`/ui/cluster/group/${values.name}`}
+              />{" "}
```
A cluster group is a defined set of cluster members, not a cluster member itself. I think we do need a different resource type for it. Maybe also a different icon?

mas-who replied:
I think in this case we can just use the cluster icon instead? Will confirm with @piperdeck

mas-who replied:
Mapped cluster groups to the `cluster-host` icon for now as discussed.

edlerd commented on src/pages/images/actions/UploadCustomIsoBtn.tsx:
```diff
@@ -20,7 +21,8 @@ const UploadCustomIsoBtn: FC<Props> = ({ className }) => {
     toastNotify.success(
       <>
-        Image <b>{name}</b> uploaded successfully
+        Image <ResourceLabel type="iso-volume" value={name} /> uploaded
```
```suggestion
        Custom ISO <ResourceLabel type="iso-volume" value={name} /> uploaded
```
I think this was a bug? Maybe check the rest of the file, it might be wrong in errors as well.

edlerd commented on src/pages/instances/CreateInstance.tsx:
```diff
@@ -172,7 +182,7 @@ const CreateInstance: FC = () => {
     toastNotify.success(
       <>
-        Created instance {instanceLink}.{message}
+        Created instance {instanceLink}. {message}
```
```suggestion
        Created instance {instanceLink}.{message}
```
There shouldn't be a dangling space here. Message might be null. The only case where it is set, it is wrapped with a p tag, so will jump into another line anyhow.

edlerd commented on src/util/instances.tsx:
```diff
@@ -1,34 +1,47 @@
 
 export const instanceLinkFromName = (args: {
   instanceName: string;
+  instanceType: LxdInstance["type"];
   project?: string;
 }): ReactNode => {
-  const { project, instanceName } = args;
+  const { project, instanceName, instanceType } = args;
   return (
-    <InstanceLink instance={{ name: instanceName, project: project || "" }} />
+    <ResourceLink
+      type={instanceType}
+      value={instanceName}
+      to={`/ui/project/${project}/instance/${instanceName}`}
+    />
   );
 };
```
I think we should continue using instance link component (see below).

edlerd commented on src/pages/instances/actions/AttachIsoBtn.tsx:
```diff
@@ -46,14 +47,19 @@ const AttachIsoBtn: FC<Props> = ({ instance }) => {
-                ISO <b>{attachedIso?.source ?? ""}</b> detached from{" "}
-                {instanceLink}
+                ISO{" "}
+                <ResourceLabel
+                  type="iso-volume"
+                  value={attachedIso?.source ?? ""}
+                />{" "}
```
Should be bold after merging the new resource label. Applies to probably all other cases where we use the label in a notification.

mas-who replied:
yes exactly, that was the first task after I rebased main :)

edlerd commented on src/util/instanceStart.tsx:
```diff
@@ -34,25 +33,27 @@ export const useInstanceStart = (instance: LxdInstance) => {
       instance.status === "Frozen" ? unfreezeInstance : startInstance;
+
+    const instanceLink = (
+      <ResourceLink
+        type={instance.type}
+        value={instance.name}
+        to={`/ui/project/${instance.project}/instance/${instance.name}`}
+      />
```
I think we should keep the old `<InstanceLink instance={instance} />` helper to avoid repeating this bit over and over again.

mas-who replied:
Yeah agreed :+1: `InstanceLink` also gets used in other UI components, would you be fine if this PR introduced bits of non-notification related changes? Alternatively, I could create another component `InstanceLinkChip`, and for other PRs I will replace `InstanceLink` with it until all updates are done?

mas-who replied:
Going with a new `InstanceLinkChip` component. I don't want to introduce non-notification related changes into this PR, and it would be simple enough to replace the components later.

edlerd replied:
Yeah, you are right, we want to keep instance link for the instance table use case anyway long term as is. Better introduce a new one.

edlerd replied:
While the helper was introduced, this bit is not using it yet. Along with a couple of other places.

edlerd reviewed (commented):
Managed to only get through ~25% of the files yet, but already found some ideas for simplification.

edlerd commented on src/pages/instances/CreateInstance.tsx:
```diff
@@ -278,6 +293,7 @@ const CreateInstance: FC = () => {
     // we still need to keep the old way of create and start instances since users may be using an older version of lxd
     const instancePayload = {
       ...instance,
+      type: values.instanceType,
```
Why add the type here? Can't we access `values.instanceType` below directly?

edlerd commented on src/pages/instances/InstanceDetailHeader.tsx:
```diff
@@ -58,6 +57,7 @@ const InstanceDetailHeader: FC<Props> = ({
           const instanceLink = instanceLinkFromName({
             instanceName: values.name,
             project,
+            instanceType: instance?.type || "container",
```
```suggestion
            instanceType: instance?.type || "container",
            project,
```
Better keep this ordered?

edlerd commented on src/pages/instances/forms/CreateImageFromInstanceSnapshotForm.tsx:
```diff
@@ -30,14 +31,17 @@ const CreateImageFromInstanceSnapshotForm: FC<Props> = ({
 
   const notifySuccess = () => {
     const created = (
       <Link to={`/ui/project/${instance.project}/images`}>created</Link>
```
Shouldn't we use the image chip with its name here?

edlerd replied:
Meaning a resource link for the image itself.

mas-who replied:
I might be wrong since I don't have the complete context for this feature, looking at the code though I think this was done before because we don't have a way to get the image name at this point?

edlerd replied:
Right, neither the form inputs nor the post response, nor the operation response contain the image description. Then let's keep it as simple as it is now.

edlerd commented on src/pages/instances/forms/CreateImageFromInstanceSnapshotForm.tsx:
```diff
@@ -30,14 +31,17 @@ const CreateImageFromInstanceSnapshotForm: FC<Props> = ({
   const toastNotify = useToastNotification();
   const queryClient = useQueryClient();
+  const snapshotLabel = (
+    <ResourceLabel bold type="snapshot" value={snapshot.name} />
```
I think this should be a chip with a link to instance detail page, snapshots section.

edlerd commented on src/pages/instances/forms/CreateInstanceSnapshotForm.tsx:
```diff
@@ -72,8 +72,15 @@ const CreateInstanceSnapshotForm: FC<Props> = ({
+                  Snapshot{" "}
+                  <ResourceLabel bold type="snapshot" value={values.name} />{" "}
+                  created for instance{" "}
+                  <ResourceLabel
```
Why not a link for instance and snapshot?

edlerd commented on src/pages/instances/forms/EditInstanceSnapshotForm.tsx:
```diff
@@ -43,8 +43,9 @@ const EditInstanceSnapshotForm: FC<Props> = ({
-        {instance.name}.
+        Snapshot <ResourceLabel bold type="snapshot" value={name} /> saved for
+        instance{" "}
+        <ResourceLabel bold type={instance.type} value={instance.name} />.
```
Better using links here.

edlerd commented on src/pages/instances/forms/UploadInstanceBackupFileForm.tsx:
```diff
@@ -60,19 +62,18 @@ const UploadInstanceBackupFileForm: FC<Props> = ({
       <>
-        Created instance <strong>{instanceName}</strong>.
+        Created instance{" "}
+        <ResourceLink type="virtual-machine" value={instanceName} to={url} />.
```
This can be a container as well

edlerd replied:
Meaning the type assumption virtual-machine is not correct, this needs to be dynamic I think.

mas-who replied:
Potentially, we can also chain an API call to a successful event and fetch the instance based on the name returned by the operation.

edlerd replied:
Yeah, I think a generic instance type would be fine here and in the other places.

edlerd commented on src/pages/instances/forms/UploadInstanceBackupFileForm.tsx:
```diff
@@ -107,7 +108,7 @@ const UploadInstanceBackupFileForm: FC<Props> = ({
           <>
             Upload completed. Now creating instance{" "}
-            <strong>{values.name}</strong>.
+            <ResourceLabel bold type="virtual-machine" value={values.name} />.
```
type can be container

mas-who replied:
will resolve with generic `instance` type here

edlerd commented on src/pages/permissions/actions/DeleteIdpGroupsModal.tsx:
```diff
@@ -25,9 +26,15 @@ const DeleteIdpGroupsModal: FC<Props> = ({ idpGroups, close }) => {
+    const successMessage = hasOneGroup ? (
+      <>
+        IDP group{" "}
+        <ResourceLabel bold type="auth-group" value={idpGroups[0].name} />{" "}
```
I think this should be a different type than auth-group.

edlerd replied:
Shall we introduce a new `idp-group` resource type?

mas-who replied:
sounds good, done. Both `auth-group` and `idp-group` still map to the same icon though.

edlerd commented on src/pages/permissions/panels/CreateGroupPanel.tsx:
We are missing the error cases, I think also in other files, we should ensure we update the error cases as well.

edlerd replied:
I think we should do the same as in src/pages/instances/CreateInstance.tsx:164 adding the chip with the link as a 3rd argument into the body.

mas-who replied:
Okay that aligns with what I suggested, that implementation concatenates the chip with the error message

edlerd replied:
![image](https://github.com/user-attachments/assets/b9a427d1-a762-4939-b82a-0ab6563da4fe)

mas-who replied:
yeah, `toastNotify.failure` uses the [messageBuilder](https://github.com/canonical/react-components/blob/d5ff954caa342ecbc0c141ed5a6d611c59cacf46/src/components/NotificationProvider/messageBuilder.tsx#L32) from react-components. The function concatenates the error message and the additional message input (in this case it would be the `instanceLink`.

edlerd commented on src/pages/projects/CreateProject.tsx:
```diff
@@ -111,7 +112,12 @@ const CreateProject: FC = () => {
-          toastNotify.success(`Project ${values.name} created.`);
+          toastNotify.success(
+            <>
+              Project <ResourceLabel bold type="project" value={values.name} />{" "}
```
Should this be a link?

mas-who replied:
I think it could be, do we link to the instance list for that project?

mas-who replied:
perfect, I'm doing a project wide scan again to see if I can convert labels into links. So this will be part of that change :+1:

edlerd reviewed (changes requested):
Managed ~50% of the changes. Some comments below, I think there are two changes needed with a bigger scope:
1. Update the error case as well, in several places only the success case was updated
2. Apply links wherever possible. There are several components, that use a label, where we should use the link component. To ease this process, we should probably create helpers similar to the `InstanceLinkChip` component.

mas-who commented:
@edlerd I've addressed the comments as well as:

1. Converted `ResourceLabel` to `ResourceLink` where appropriate.
2. Added links/labels to error messages, but only for the toast notifications. Thought about adding links to inline notifications, but I think that is not necessary since inline notifications are directly related to the context of a resource's detail page.

edlerd commented on src/util/operations.tsx:
```diff
@@ -13,6 +13,26 @@ export const getInstanceName = (operation?: LxdOperation): string => {
   );
 };
 
+export const getSnapshotName = (operation?: LxdOperation): string => {
+  // the url can be one of below formats
+  // instance and volume snapshot operations are mutually exclusive
+  // /1.0/instances/<instance_name>/snapshots/<snapshot_name>
+  // /1.0/storage-pools/<pool_name>/volumes/custom/<volume_name>/snapshots/<snapshot_name>
+  const instanceSnapshots = operation?.resources?.instances_snapshots ?? [];
+  const storageVolumeSnapshots =
+    operation?.resources?.storage_volume_snapshots ?? [];
+
+  if (instanceSnapshots.length) {
+    return instanceSnapshots[0].split("/")[5].split("?")[0];
+  }
+
+  if (storageVolumeSnapshots.length) {
+    return storageVolumeSnapshots[0].split("/")[8].split("?")[0];
+  }
+
+  return "";
+};
+
```
please add a spec test for this one, similar to the other functions in this file.

edlerd replied:
Also, it might be better to split this by instance and volume snapshots. Because the logic has no relevant overlap and the use cases are distinct.

mas-who replied:
sounds good, will do both. Let's stay consistent with the pattern to have separate functions/components for instance and volume snapshots

edlerd commented on src/pages/projects/ProjectConfigurationHeader.tsx:
```diff
@@ -47,28 +48,44 @@ const ProjectConfigurationHeader: FC<Props> = ({ project }) => {
+        <ResourceLink
+          type="project"
+          value={values.name}
+          to={`/ui/project/${project.name}/configuration`}
```
This is confusing, because it is linking to the old url. Maybe give the variable holding it a better name?

mas-who replied:
yeah you are right, calling it `oldProjectLink`.

mas-who commented on src/pages/instances/SnapshotLinkChip.tsx:
great, done :+1:

mas-who commented on src/types/helpers.d.ts:
thanks this is good advice! :+1:

edlerd reviewed (changes requested):
Reviewed all the files, some concerns below.

mas-who commented on src/pages/storage/forms/EditStorageVolume.tsx:
```diff
@@ -62,7 +63,17 @@ const EditStorageVolume: FC<Props> = ({ volume }) => {
+            <>
+              Storage volume{" "}
+              <ResourceLink
+                type="profile"
```
Thanks for the spot, I did a full check across the entire codebase again, seems other places are fine.

mas-who commented on src/types/instance.d.ts:
```diff
@@ -102,6 +102,6 @@ export interface LxdInstance {
   stateful: boolean;
   status: LxdInstanceStatus;
-  type: string;
+  type: "container" | "virtual-machine" | "instance";
```
removed, instead, I created a new `InstanceIconType` that gets used for these function definitions

mas-who commented on src/util/instances.tsx:
```diff
@@ -1,34 +1,30 @@
-  return <InstanceLink instance={{ name: linkText, project: project || "" }} />;
+  return (
+    <InstanceLinkChip
+      instance={{ name: linkText, project: project || "", type: instanceType }}
```
fixed both places to default to "default"

edlerd commented on src/pages/instances/forms/DuplicateInstanceForm.tsx:
```diff
@@ -64,20 +66,22 @@ const DuplicateInstanceForm: FC<Props> = ({ instance, close }) => {
+    type: InstanceIconType,
+  ) => {
+    const url = `/ui/project/${project}/instance/${name}/configuration`;
     const message = (
       <>
-        Created instance <strong>{instanceName}</strong>.
+        Created instance <ResourceLink type={type} value={name} to={url} />.
```
The chip should direct to the overview section. Maybe check other places, there might be more outliers like this one where we reuse the url.

mas-who replied:
I remember this was discussed when we did the duplicate features, since the user can't configure the instance in the duplicate form, we want to create a notification that directs them to the config page. This is also indicated by the "Configure" action associated to the notification. I think that still makes sense and I'm not sure if we should now redirect to overview page?

edlerd replied:
I am fine with the configure link going to the configure page. But I think the chips should always go to the overview page, so the user is confirmed in the mental modal of how the chips work.

edlerd commented on src/pages/storage/forms/DuplicateVolumeForm.tsx:
```diff
@@ -53,17 +53,18 @@ const DuplicateVolumeForm: FC<Props> = ({ volume, close }) => {
 
   const notifySuccess = (volumeName: string, project: string, pool: string) => {
+    const url = `/ui/project/${project}/storage/pool/${pool}/volumes/custom/${volumeName}/configuration`;
     const message = (
       <>
-        Created volume <strong>{volumeName}</strong>.
+        Created volume{" "}
+        <ResourceLink type="volume" value={volumeName} to={url} />.
```
Here we also link the config tab, where the target url for the chip should be the overview page.

mas-who replied:
resolved with the previous comment

edlerd reviewed (changes requested):
Probably one last concern found below.

edlerd reviewed (approved):
LGTM, thanks for the many iterations to drive this to perfection!
